android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/BroadcastActivity.java

```diff
@@ -0,0 +1,205 @@
+                            Toast.LENGTH_LONG).show();
+                    }
+                };
+                UrlShortenerClient.getInstance(this).shortenUrl(url, urlSetter, TAG);
+            }
+        } else {
+            Toast.makeText(this, getString(R.string.no_url_error),
+                Toast.LENGTH_LONG).show();
+            finish();
+        }
+    }
+
+    private void checkBleAndStart(){
+        if (checkAndWarnBleAdvertiseCapability()) {
+            parseAndHandleUrl();
+        }
+    }
+    private boolean checkAndWarnBleAdvertiseCapability() {
+        if (!hasBleAdvertiseCapability()) {
+            Toast.makeText(this, getString(R.string.ble_error),
+                Toast.LENGTH_LONG).show();
+            return false;
+        }
+
+        return true;
+    }
+
+    // Check if bluetooth is on
+    private boolean checkIfBluetoothIsEnabled() {
+        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(
+            Context.BLUETOOTH_SERVICE);
+        BluetoothAdapter bluetoothAdapter = bluetoothManager.getAdapter();
+        if (bluetoothAdapter == null || !bluetoothAdapter.isEnabled()) {
+          Log.d(TAG, "not enabled");
+            return false;
+        }
+        Log.d(TAG, "enabled");
+        return true;
+    }
+
+    // Ask the user to enable bluetooth
+    private void requestBluetooth() {
+        Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
+        startActivityForResult(enableBtIntent, ENABLE_BLUETOOTH_REQUEST_ID);
+    }
+
+    // Check if the given bluetooth hardware
+    // on the current device supports ble advertisements
+    @TargetApi(21)
+    private boolean hasBleAdvertiseCapability() {
+        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(
+            Context.BLUETOOTH_SERVICE);
+        if (bluetoothManager.getAdapter().getBluetoothLeAdvertiser() == null) {
+            Log.d(TAG, "cant advertise");
+            return false;
+        }
+        Log.d(TAG, "can advertise");
+        return true;
+    }
+
+    private void startBroadcastService(String displayUrl, String url){
+        Intent intent = new Intent(this, PhysicalWebBroadcastService.class);
+        intent.putExtra(PhysicalWebBroadcastService.BROADCAST_URL_KEY, url);
+        intent.putExtra(PhysicalWebBroadcastService.DISPLAY_URL_KEY, displayUrl);
+        startService(intent);
+    }
+
+
+    // Check if the given text contains a URL
+    private String findUrlInText(String text) {
+        List<String> urls = new ArrayList<>();
+        Matcher m = Patterns.WEB_URL.matcher(text);
+        while (m.find()) {
+            String url = m.group();
+            urls.add(url);
+        }
+        if (urls.size() > 0) {
+            return urls.get(0);
+        }
+
+        return null;
+    }
+
+    private static boolean hasValidUrlLength(String url) {
+        int uriLength = UriBeacon.uriLength(url);
```

Comment:
Let's try to remove the dependency on UriBeacon.  The implementation of uriLength is here:

https://github.com/google/uribeacon/blob/uribeacon-final/android-uribeacon/uribeacon-library/src/main/java/org/uribeacon/beacon/UriBeacon.java#L227

Please add it to AdvertiseDataUtils.